packages/server/lib/controllers/ratelimit.middleware.ts

```diff
@@ -0,0 +1,70 @@
+import type { Request, Response, NextFunction } from 'express';
+import { createClient } from 'redis';
+import { RateLimiterRedis, RateLimiterMemory, RateLimiterRes } from 'rate-limiter-flexible';
+import { getAccount, getRedisUrl } from '@nangohq/shared';
+import { logger } from '@nangohq/shared';
+
+const rateLimiter = await (async () => {
+    const opts = {
+        keyPrefix: 'middleware',
+        points: 1200,
+        duration: 60,
+        blockDuration: 0
+    };
+    const url = getRedisUrl();
+    if (url) {
+        const redisClient = await createClient({ url: url, disableOfflineQueue: true }).connect();
+        redisClient.on('error', (err) => {
+            logger.error(`Redis (rate-limiter) error: ${err}`);
+        });
+        return new RateLimiterRedis({
+            storeClient: redisClient,
+            ...opts
+        });
+    }
+    return new RateLimiterMemory(opts);
+})();
+
+export const rateLimiterMiddleware = (req: Request, res: Response, next: NextFunction) => {
+    const setXRateLimitHeaders = (rateLimiterRes: RateLimiterRes) => {
+        const resetEpoch = Math.floor(new Date(Date.now() + rateLimiterRes.msBeforeNext).getTime() / 1000);
+        res.setHeader('X-RateLimit-Limit', rateLimiter.points);
+        res.setHeader('X-RateLimit-Remaining', rateLimiterRes.remainingPoints);
+        res.setHeader('X-RateLimit-Reset', resetEpoch);
+    };
+    const key = getKey(req, res);
+    const pointsToConsume = getPointsToConsume(req);
+    rateLimiter
+        .consume(key, pointsToConsume)
+        .then((rateLimiterRes) => {
+            setXRateLimitHeaders(rateLimiterRes);
+            next();
+        })
+        .catch((rateLimiterRes) => {
+            res.setHeader('Retry-After', Math.floor(rateLimiterRes.msBeforeNext / 1000));
+            setXRateLimitHeaders(rateLimiterRes);
+            logger.info(`Rate limit exceeded for ${key}. Request: ${req.method} ${req.path})`);
+            next();
+            // TODO:
+            // res.status(429).send('Too Many Requests');
```

Comment:
As a first step, we are just logging and not sending the 429. Just to make sure there is no obvious issue. I will let it run a few hours like this before to start sending 429